generator/src/main/java/org/corfudb/generator/Generator.java

```diff
@@ -19,55 +19,201 @@
+                            op.execute();
+                        } catch (Exception e) {
+                            log.error("Operation failed with", e);
+                        }
+                    }
+                }
+            });
+        }
+    }
 
+    /**
+     * A thread is tasked with checkpoint and trim activity. This is a cyclic operation.
+     */
+    private void runCpTrimOperation() {
         Runnable cpTrimTask = () -> {
             Operation op = new CheckpointOperation(state);
             op.execute();
         };
-
-        final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
         scheduler.scheduleAtFixedRate(cpTrimTask, 30, cpPeriod, TimeUnit.SECONDS);
+    }
 
-        ExecutorService appWorkers = newWorkStealingPool(numThreads);
 
-        Runnable app = () -> {
-            List<Operation> operations = state.getOperations().sample(numOperations);
-            for (Operation operation : operations) {
-                try {
-                    operation.execute();
-                } catch (Exception ex) {
-                    throw new RuntimeException(ex);
+    public void shutdown() {
+        workers.shutdown();
+        scheduler.shutdown();
+    }
+
+
+    /**
+     * Run a Longevity test that 'freezes' the system every CHUNK_RUNTIME,
+     * sends the correctness log (of one chunk) to verification, then resumes
+     * the run.
+     * The same action happens in a cyclic fashion until the totalRuntime.
+     */
+    public void runLongevityTestForVerification() {
+
+        fillOperationQueue();
+        long lastPollTime = System.currentTimeMillis();
+        consumeOperationQueue();
+        runCpTrimOperation();
+
+        while (true) {
+            // One round of 'chunk' starts
+            long cumulativeRuntime = 0;
+            while (true) {
+                Sleep.sleepUninterruptibly(POLL_TIME);
+
+                // Check if the operationQueue is drained or not,
+                // if it's empty and has not reached CHUNK_RUNTIME,
+                // then fill up the queue for further consuming.
+                if (operationQueue.isEmpty()) {
+                    // Calculate the 'cumulative' runtime so far,
+                    // skipping the time for filling up the operation queue
+                    cumulativeRuntime += System.currentTimeMillis() - lastPollTime;
+                    // Break this while loop if it reaches CHUNK_RUNTIME
+                    if (cumulativeRuntime >= CHUNK_RUNTIME.toMillis()) {
+                        // At the end of the chunk, give some margin
+                        // for the workers to finish their operations
+                        Sleep.sleepUninterruptibly(CHUNK_TIMEOUT);
+                        Correctness.recordOperation("Liveness, " + "Success", false);
+                        break;
+                    }
+                    fillOperationQueue();
+                    lastPollTime = System.currentTimeMillis();
```

Comment:
I was thinking about your comment on`operationQueueLock` in my first review and I found there is a race condition here. Imagine there is only one operation left in the queue, and more than one consumer thread have passed the `if (!operationQueueLock && !operationQueue.isEmpty())` check, then only one thread can take the only operation in the queue, and all the other threads will block on `operationQueue.take()`. After some time, `fillOperationQueue` starts and adding operations one by one, then some threads blocking on `take()` will consume the operations just added, even if `fillOperationQueue()` is not finished since `if (!operationQueueLock...)` check is already done before the thread is blocked. This race condition leads to the wrong calculation of totalRuntime.

Reply:
yes, we need to fix this. will discuss with you offline.

Reply:
At second thought, I don't think what you describe will happen. will explain to you offline.

Reply:
anyway, will discuss this offline.